# Singly Linked List

Singly Linked List consists of nodes, and each node consists of data and address of next node. In this program let's look into basic algorithm on linked lists, that is adding an element in beggining and in the ending, remove a particular element by value and printing the linked list

The class Node below consists of data and nextnode. The value of nextnode is set to None by default since we will only have data in it when we initialize a node, and it will be an independent node.

To remove an element in Linked List we need to use the node class, as the address of previous and next nodes are traced by nodes itself. If the removing data is present in the very first node then it is deleted in the O(1) time complexity, other we have to traverse through all the nodes to find the element and delete it

In [40]:
class Node(object):
    def __init__(self, data):
        self.data = data
        self.nextNode = None
    def remove(self, data, previousNode):
        if(self.data == data):
            previousNode.nextNode = self.nextNode
            del self.data
            del self.nextNode
        elif self.nextNode is not None: 
            self.nextNode.remove(data, self)
        else: print("there is no such data")

By default, the linked list class has only one head, which points nowhere/None because it is alone. The head is pointing towards the first node when it is produced, and the first node is pointing nowhere/to none until the next node is added.

To add the element at the start of the list, we create a basic node and connect it to the next node if it exits; otherwise, we simply create a node, then make the head reference to this newly created node, and the element is added at the start of the list. This is done in O(1) time complexity as we are just creating a new node and doing few simplel operations to finish the adding

To add the element at the end of the list, we must first traverse the list until we reach the end, and then add the newly created node by having the last node point to the new node. This can be done in O(n) time complexity as we cannot directly go to last element and we have to traverse through all elements

While finding the size of the linked list we are creating a count variable and then using it as a counter while traversing the list, this will take the O(n) complexity as here again we are traversing the list. But this complexity can possibly be reduced to O(1) by maintaining the counter variable in innit itself and increasing or decreasing the value everytime the operation is done on the list.

Removing an element in list is also a O(n) complexity as we are finding the element throuhg traversing the list and if we do not find the element after traversal in our worst case then we are not removing the element.

In [41]:
class LinkedList(object):
    def __init__(self):
        self.head = None
        
    #O(1)
    def insertFirst(self, data):
        self.counter += 1
        newNode = Node(data)
        if self.head is None:
            self.head = newNode
        else:
            newNode.nextNode = self.head
            self.head = newNode
    
    #O(n)
    def insertLast(self, data):
        newNode = Node(data)
        if self.head is None:
            self.head = newNode
        else:
            actualNode = self.head
            while actualNode.nextNode is not None:
                actualNode = actualNode.nextNode
            actualNode.nextNode = newNode
    
    #O(n)
    def sizeOfList(self):
        count = 1
        actualNode = self.head
        while actualNode.nextNode is not None:
            count += 1
            actualNode = actualNode.nextNode
        return count
    
    #O(n)
    def remove(self, data):
        if self.head is None:
            print("there is no data to be removed")
        elif data == self.head.data:
            self.head = self.head.nextNode
        else:
            self.head.nextNode.remove(data, self.head)
    
    #O(n)
    def traverse(self):
        actualNode = self.head
        pos = 0
        while actualNode is not None:
            pos += 1
            print(f"the element at {pos} position is {actualNode.data}" )
            actualNode = actualNode.nextNode 

In [42]:
linkedList = LinkedList();
linkedList.insertLast(12);
linkedList.insertLast(122);
linkedList.insertLast(1212);
linkedList.insertFirst(112);

linkedList.traverse()
print(f"the size of the array is {linkedList.sizeOfList()}")
linkedList.remove(112)
print(f"after removing element 112")

linkedList.traverse()

print(f"the size of the array is {linkedList.sizeOfList()}")

the element at 1 position is 112
the element at 2 position is 12
the element at 3 position is 122
the element at 4 position is 1212
the size of the array is 4
after removing element 112
the element at 1 position is 12
the element at 2 position is 122
the element at 3 position is 1212
the size of the array is 3


## I. MATHEMATICAL PROBLEMS

## Check if number is plaindrome

In [3]:
def checkPalindrome(val):
    if not val//10: #checking for single digit number
        return True
    else:
        rev = 0 #reverse number initiation
        actual = val #storing actual number to compare later
        while val: #in each iteration we are removing one digit, running loop till we have numeber
            temp = val%10 #getting last digit
            rev = rev*10 + temp #making reverse numbers
            val = val//10 #removing last digit
        return True if actual == rev else False #returning value

In [4]:
print(checkPalindrome(303))

True


## Factorial of N

Factorial is defined as multiplication of all numbers till that number and N>0.

For example factorial of 4 is 1 * 2 * 3 * 4 = 24 and factorial of 6 is 1 * 2 * 3 * 4 * 5 * 6 = 720

In [7]:
def factorial(num):
    val = 1
    for i in range(1, num+1):
        val = val*i
    return val

In [9]:
factorial(4)

24

In [10]:
factorial(6)

720

## Factorial using recursion

fact(4) calls:

    4 * fact(3)
    
    3 * fact(2)
    
    2 * fact(1)
    
    1 * fact(0)
    
fact(0) returns 1

and then the stack return 1 * 2 * 3 * 4 * 5 * 6
    
    
In this case the function take theta(n) space as we will have a point where there will be n+1 function calls in a stack, whereas in above case we had a function with constant space.


In [21]:
def factorials(num):
    if(num < 0): return 0
    if(num == 0):
        return 1
    else: return num*factorials(num - 1)

In [22]:
factorials(5)

120

## Trailing Zeros in factorial 

we are given a number and our task is to find trailing zeros in factorial of that number

n = 5

factorial = 120

op = 1

In [48]:
def trailingZeros(num):
    fact = 1
    count = 0
    for i in range(1, num+1):
        fact = fact*i
    print(fact)
    while not fact%10:
        count += 1
        fact = fact//10
    return count

In [53]:
trailingZeros(5)

120


1

In [51]:
trailingZeros(20)

2432902008176640000


4

The above solution will run good for all the values. But as the value gets bigger, the factorial number gets even larger. In python we might not face any issue but this logic will give overflow issues in other languages, also its not the most efficient way of handling space complexity

we can avoid this by finding a pair of 2 * 5 as this is what will lead to a trailing zero. To make this even simpler we can count number of 5's we have as in most cases number of 5's will be less.

1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * 10 ......

In prime factorization of number we can see that we have 5 repeating for atleast 5 values, that means we can safely assume that we have floor of (n/5) 5's in any number. i.e 5 is repeating for each 5 numbers atleast



In [79]:
def trailingZeros2(num):
    count = 0
    i = 5
    while i <= num:
        count = count + num//i
        i *= 5
    return count

In [82]:
trailingZeros2(251)

62

In [83]:
trailingZeros(251)

811446921488186040812524452558116486219705773136118947353733549205318760636332913581693296758128078062183055345650547664609463859915056696520548600046891762340808337287505448037728337751512715064266648553413711482053864096073935168807868779389231199415987090547371213614453678579719100407204882954963078533417133568165939310960491560595435373673195910930411701938747154349882416287042721093688617485205308997682124247612021055212748800000000000000000000000000000000000000000000000000000000000000


62

## Greatest Common Divisor

we need to find a greatest common divisor between two numbers

example, a = 4, b = 6

o/p - 2

two numbers does not have a common divisor except 1 then it can be composit as well as prime. Example 3, 4 and 9, 10

relation of GCD with tiling problem.

make rectangle of a X b and finding sides of the largest square using which we can tile the whole rectangle.

## approach

start with assuming smallest number is a gcd and common divisor cannot me more than the lowest number

start decrementing the this number and see if both numbers are divisible by this gcd

in case it matches that is gcd and break the loop.

in worst case the value is 1.

In [15]:
def gcd(a, b):
    limit = a if a<b else b
    gcd = limit
    for i in range(1, limit+1)[::-1]:
        if(a%i == 0 and b%i ==0):
            gcd = i
            break
        else: continue
    return gcd

In [16]:
gcd(7, 13)

1

In [17]:
gcd(100, 200)

100

## euclidean algorithm for GCD

basic idea is that is b is smaller than a then gcd(a, b) = gcd(a-b, b)

why? lets write a theory. Let g be the gcd of a and b

a = gx, b = gy and gcd(x, y) = 1

(a-b) = g(x-y)

Example go through

a = 12, 15

12, 3

9, 3

6, 3

3, 3

In [21]:
def euclideanGCD(a, b):
    while(a!=b):
        if a>b:
            a = a-b
        else: b = b-a
        
    return a

In [22]:
euclideanGCD(100, 200)

100

In [23]:
euclideanGCD(2, 3)

1

## optimised euclidean algorithm

## walkthrough

12, 15

15, 12

12, 3

3, 0

In [25]:
def euclideanGCD(a, b):
    if(b == 0):
        return a
    else: return euclideanGCD(b, a%b)

In [26]:
euclideanGCD(100, 200)

100

In [28]:
def euclideanGCD(a, b):
    return a if b==0 else euclideanGCD(b, a%b)

In [29]:
euclideanGCD(100, 200)

100

## LCM of two numbers

Least common multiple, the smallest number divisible by both numbers

a, b = 4, 6

o/p = 12

there are other multiple of 4, 6 such as 12, 24, 36, etc but the least commong factor is 12

a, b = 2, 8

o/p = 8

when one number is divisible by another then LCM is largest number

a, b = 2, 3

o/p = 6

If two numbers are co prime then LCM is multiplication of those two numbers

In [6]:
# naive approach
def lcm(a, b):
    maxlcm = a if a>=b else b
    while(True):
        if(maxlcm % a == 0 and maxlcm % b ==0):
            return maxlcm
        else: maxlcm += 1
    return maxlcm    

In [7]:
lcm(3, 4)

12

In [9]:
lcm(2, 8)

8

In [10]:
lcm(4, 6)

12

## Formula for LCM

a * b = gcd(a, b) * lcm(a, b)

In [12]:
def lcm2(a, b):
    def gcd(a, b):
        return a if b==0 else gcd(b, a%b)

    return (a*b)//gcd(a, b)

In [13]:
lcm2(2, 3)

6

In [14]:
lcm2(4, 6)

12

# Check if number is prime

a number is prime if it is divisible by 1 and itself

In [23]:
def isPrime(a):
    return False if a%i == 0 for i in range(a) else True

SyntaxError: invalid syntax (<ipython-input-23-637bee6f1099>, line 2)